src/v/datalake/translation/partition_translator.cc

```diff
@@ -0,0 +1,348 @@
+                        e);
+                  }
+              });
+        });
+    });
+    vlog(_logger.debug, "started partition translator in term {}", _term);
+}
+
+std::chrono::milliseconds partition_translator::translation_interval() const {
+    return _jitter.base_duration();
+}
+
+void partition_translator::reset_translation_interval(
+  std::chrono::milliseconds new_base) {
+    _jitter = jitter_t{new_base, translation_jitter};
+    vlog(
+      _logger.info,
+      "Iceberg translation interval reset to: {}",
+      _jitter.base_duration());
+}
+
+ss::future<> partition_translator::stop() {
+    vlog(_logger.debug, "stopping partition translator in term {}", _term);
+    auto f = _gate.close();
+    _as.request_abort();
+    co_await std::move(f);
+}
+
+kafka::offset partition_translator::min_offset_for_translation() const {
+    return model::offset_cast(_partition_proxy->local_start_offset());
+}
+
+std::optional<kafka::offset>
+partition_translator::max_offset_for_translation() const {
+    // We factor in LSO to ensure only committed transactional batches are
+    // translated.
+    auto lso = _partition_proxy->last_stable_offset();
+    if (lso.has_error()) {
+        vlog(
+          _logger.warn,
+          "unable to compute lso for translation: {}",
+          lso.error());
+        return std::nullopt;
+    }
+    return kafka::prev_offset(model::offset_cast(lso.value()));
+}
+
+ss::future<std::optional<coordinator::translated_offset_range>>
+partition_translator::do_translation_for_range(
+  retry_chain_node& parent, model::record_batch_reader rdr) {
+    // This configuration only writes a single row group per file but we limit
+    // the bytes via the reader max_bytes.
+    auto writer_factory = std::make_unique<batching_parquet_writer_factory>(
+      local_path{_writer_scratch_space},
+      "", // todo(iceberg): generate a prefix with offset information
+      max_rows_per_row_group,
+      max_bytes_per_row_group);
+    auto task = translation_task{**_cloud_io};
+    const auto& ntp = _partition->ntp();
+    auto remote_path_prefix = remote_path{
+      fmt::format("{}/{}/{}", iceberg_file_path_prefix, ntp.path(), _term)};
+    lazy_abort_source las{[this] {
+        return can_continue() ? std::nullopt
+                              : std::make_optional("translator stopping");
+    }};
+    auto result = co_await task.translate(
+      std::move(writer_factory),
+      std::move(rdr),
+      remote_path_prefix,
+      parent,
+      las);
+    if (result.has_error()) {
+        vlog(_logger.warn, "Error translating range {}", result.error());
+        co_return std::nullopt;
+    }
+    co_return std::move(result.value());
+}
+
+ss::future<partition_translator::translation_success>
+partition_translator::do_translate_once(retry_chain_node& parent_rcn) {
+    if (
+      !_partition->get_ntp_config().iceberg_enabled()
+      || !_features->local().is_active(features::feature::datalake_iceberg)) {
```

Comment:
I think the following scenario is possible.

- 3 node cluster, mixed mode, 1 node is upgraded to RP version with iceberg support and hosts the controller leader
- User creates an iceberg enabled topic and the upgrade node also has data partition leader.
- If we do not have this check in place, we end up replicating the "translation_state" batch to the followers on a lower version.
- If we only do this check in datalake manger and if the feature is not enabled by the time a translator instance is created and the leader never moves from the node, we risk missing a notification unless we periodically check for it (which is what this logic does).

Reply:
But presumably we'd prevent iceberg topics from being created until the feature says that the whole cluster supports it, no? So step 2 would be blocked?

Reply:
I don't think we do (not just for iceberg but for any feature), my guess is it is generally preferred to push the mixed mode complexity (wherever possible) into the feature itself rather than erroring out client requests until the upgrade finishes.

Reply:
That's very interesting. Especially for net new features, I think it will easier to reason about the entire feature if we gate as early as possible, because reasoning about it downstream leaves around a ton of ambiguous edge cases (e.g. why gate here instead of in the datalake manager? are we going to create the coordinator in mixed mode? etc).

Anyway, not blocking -- we can always revisit this outside of this PR

Reply:
right definitely easier to reason about if we don't let the topics into the system.